File: core/src/org/rexcellentgames/burningknight/mod/ModManager.kt
package org.rexcellentgames.burningknight.mod

import com.badlogic.gdx.Gdx
import org.rexcellentgames.burningknight.util.Log
import java.util.*

object ModManager {
  private const val modDirectory = "mods/"
  val mods = ArrayList<Mod>()
  private var inited = false

  fun load() {
    if (inited) {
      return
    }

    val dir = Gdx.files.internal(modDirectory)

    if (!dir.exists()) {
      Log.error("Mods directory does not exist!")
      return
    }

    val files = dir.list()

    for (file in files) {
      if (file.isDirectory) {
        val mod = Mod.load(file)

        if (mod != null) {
          mods.add(mod)
          
          mod.init()
        }
      }
    }

    inited = true
  }

  fun update(dt: Float) {
    if (!inited) {
      return
    }

    for (mod in mods) {
      mod.update(dt)
    }
  }

  fun draw() {
    if (!inited) {
      return
    }

    for (mod in mods) {
      mod.draw()
    }
  }

  fun destroy() {
    if (!inited) {
      return
    }

    for (mod in mods) {
      mod.destroy()
    }

    mods.clear()
  }
}

File: BurningKnight/Assets/Mods/ModManager.cs
using System.Collections.Generic;
using BurningKnight.Util.Files;

namespace BurningKnight.Assets.Mods
{
	public class ModManager : AssetManager
	{
		private List<Mod> mods = new List<Mod>();
		
		public override void LoadAssets()
		{
			FileHandle mods = FileHandle.FromRoot("Mods/");

			if (!mods.Exists())
			{
				mods.MakeDirectory();
			}
			
			foreach (var file in mods.ListDirectoryHandles())
			{
				ParseDirectory(file);
			}
		}

		public override void Destroy()
		{
			base.Destroy();
			
			foreach (var mod in mods)
			{
				mod.Destroy();
			}
		}

		private void ParseDirectory(FileHandle dir)
		{
			Mod mod = new Mod(dir);
			mods.Add(mod);
			
			mod.Init();
		}

		public void Update(float dt)
		{
			foreach (var mod in mods)
			{
				mod.Update(dt);
			}
		}

		public void Draw()
		{
			foreach (var mod in mods)
			{
				mod.Draw();
			}
		}
	}
}